File: apps/backend/src/modules/strategy/dsl/indicators.ts
/**
 * Deterministic technical-indicator library for the strategy DSL (Phase 1).
 *
 * Every function returns an array aligned 1:1 with the input candles; positions before an
 * indicator has enough history to be defined hold `NaN` (the interpreter treats any condition
 * touching a NaN as false, and never enters before the tree's warmup bar). These formulas are
 * the CANONICAL definitions — the Python port in worker/strategy/dsl/indicators.py must match
 * them exactly, and tools/parity_evaluator.py asserts it. Keep them simple and explicit for
 * that reason (no clever vectorization that could drift between languages).
 *
 * Conventions chosen for cross-language reproducibility:
 *   - EMA seeds ema[0] = price[0] (same as the pre-DSL evaluator), k = 2/(period+1).
 *   - SMA/Bollinger use a population standard deviation (divide by N, not N-1).
 *   - RSI/ATR use Wilder smoothing seeded by a simple average of the first `period` values.
 *   - Donchian/rollingHigh/rollingLow INCLUDE the current bar in the window.
 */

export interface Series {
  open: number[];
  high: number[];
  low: number[];
  close: number[];
  volume: number[];
}

const nan = (n: number): number[] => new Array(n).fill(NaN);

export function ema(prices: number[], period: number): number[] {
  const out = new Array(prices.length).fill(0);
  if (prices.length === 0) return out;
  const k = 2 / (period + 1);
  out[0] = prices[0];
  for (let i = 1; i < prices.length; i++) out[i] = prices[i] * k + out[i - 1] * (1 - k);
  return out;
}

// NaN-safe window mean: returns NaN until there are `period` consecutive finite values ending
// at i. A sliding-sum would be poisoned forever by the NaN warmup-prefix of a source indicator
// (e.g. Stochastic smoothing the raw-%K series), so each window is summed directly.
export function sma(prices: number[], period: number): number[] {
  const out = nan(prices.length);
  for (let i = period - 1; i < prices.length; i++) {
    let sum = 0;
    let ok = true;
    for (let j = i - period + 1; j <= i; j++) {
      if (!Number.isFinite(prices[j])) {
        ok = false;
        break;
      }
      sum += prices[j];
    }
    if (ok) out[i] = sum / period;
  }
  return out;
}

/** Population standard deviation over a trailing window of `period` values. */
function rollingStd(prices: number[], period: number, means: number[]): number[] {
  const out = nan(prices.length);
  for (let i = period - 1; i < prices.length; i++) {
    const mean = means[i];
    let acc = 0;
    for (let j = i - period + 1; j <= i; j++) acc += (prices[j] - mean) ** 2;
    out[i] = Math.sqrt(acc / period);
  }
  return out;
}

export function rsi(prices: number[], period: number): number[] {
  const out = nan(prices.length);
  if (prices.length <= period) return out;
  let avgGain = 0;
  let avgLoss = 0;
  // Seed with the simple average of the first `period` changes.
  for (let i = 1; i <= period; i++) {
    const change = prices[i] - prices[i - 1];
    if (change >= 0) avgGain += change;
    else avgLoss -= change;
  }
  avgGain /= period;
  avgLoss /= period;
  out[period] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);
  // Wilder smoothing thereafter.
  for (let i = period + 1; i < prices.length; i++) {
    const change = prices[i] - prices[i - 1];
    const gain = change > 0 ? change : 0;
    const loss = change < 0 ? -change : 0;
    avgGain = (avgGain * (period - 1) + gain) / period;
    avgLoss = (avgLoss * (period - 1) + loss) / period;
    out[i] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);
  }
  return out;
}

export function atr(s: Series, period: number): number[] {
  const n = s.close.length;
  const out = nan(n);
  if (n <= period) return out;
  const tr = new Array(n).fill(0);
  tr[0] = s.high[0] - s.low[0];
  for (let i = 1; i < n; i++) {
    tr[i] = Math.max(
      s.high[i] - s.low[i],
      Math.abs(s.high[i] - s.close[i - 1]),
      Math.abs(s.low[i] - s.close[i - 1]),
    );
  }
  // Seed ATR at index `period` with the simple average of TR[1..period], then Wilder-smooth.
  let sum = 0;
  for (let i = 1; i <= period; i++) sum += tr[i];
  out[period] = sum / period;
  for (let i = period + 1; i < n; i++) out[i] = (out[i - 1] * (period - 1) + tr[i]) / period;
  return out;
}

export interface MacdResult {
  line: number[];
  signal: number[];
  hist: number[];
}
export function macd(prices: number[], fast: number, slow: number, signalPeriod: number): MacdResult {
  const emaFast = ema(prices, fast);
  const emaSlow = ema(prices, slow);
  const line = prices.map((_, i) => emaFast[i] - emaSlow[i]);
  const signal = ema(line, signalPeriod);
  const hist = line.map((v, i) => v - signal[i]);
  return { line, signal, hist };
}

export interface BandResult {
  upper: number[];
  mid: number[];
  lower: number[];
}
export function bollinger(prices: number[], period: number, mult: number): BandResult {
  const mid = sma(prices, period);
  const sd = rollingStd(prices, period, mid);
  const upper = mid.map((m, i) => m + mult * sd[i]);
  const lower = mid.map((m, i) => m - mult * sd[i]);
  return { upper, mid, lower };
}

export function rollingMax(values: number[], period: number): number[] {
  const out = nan(values.length);
  for (let i = period - 1; i < values.length; i++) {
    let m = -Infinity;
    for (let j = i - period + 1; j <= i; j++) if (values[j] > m) m = values[j];
    out[i] = m;
  }
  return out;
}

export function rollingMin(values: number[], period: number): number[] {
  const out = nan(values.length);
  for (let i = period - 1; i < values.length; i++) {
    let m = Infinity;
    for (let j = i - period + 1; j <= i; j++) if (values[j] < m) m = values[j];
    out[i] = m;
  }
  return out;
}

export interface DonchianResult {
  upper: number[];
  lower: number[];
}
export function donchian(s: Series, period: number): DonchianResult {
  return { upper: rollingMax(s.high, period), lower: rollingMin(s.low, period) };
}

export interface StochResult {
  k: number[];
  d: number[];
}
export function stochastic(s: Series, period: number, smoothK: number, smoothD: number): StochResult {
  const n = s.close.length;
  const hh = rollingMax(s.high, period);
  const ll = rollingMin(s.low, period);
  const rawK = nan(n);
  for (let i = period - 1; i < n; i++) {
    const range = hh[i] - ll[i];
    rawK[i] = range === 0 ? 0 : ((s.close[i] - ll[i]) / range) * 100;
  }
  const k = sma(rawK, smoothK);
  const d = sma(k, smoothD);
  return { k, d };
}

File: apps/worker-py/worker/strategy/dsl/indicators.py
"""Deterministic technical-indicator library for the strategy DSL.

Originally a parity port of a TypeScript twin, but the backend engine was deleted in the
single-engine consolidation (Phase B) — this is now the ONE implementation, so new indicators
(e.g. adx) live here only.

Series is a dict of equal-length float lists: {"open","high","low","close","volume"}.
Positions before an indicator is defined hold float('nan'); the interpreter treats any condition
touching a NaN as false.
"""
from __future__ import annotations

import math

NAN = float("nan")


def _nan(n: int) -> list[float]:
    return [NAN] * n


def ema(prices: list[float], period: int) -> list[float]:
    out = [0.0] * len(prices)
    if not prices:
        return out
    k = 2 / (period + 1)
    out[0] = prices[0]
    for i in range(1, len(prices)):
        out[i] = prices[i] * k + out[i - 1] * (1 - k)
    return out


# NaN-safe window mean: returns NaN until there are `period` consecutive finite values ending
# at i. A sliding-sum would be poisoned forever by the NaN warmup-prefix of a source indicator
# (e.g. Stochastic smoothing the raw-%K series), so each window is summed directly.
def sma(prices: list[float], period: int) -> list[float]:
    out = _nan(len(prices))
    for i in range(period - 1, len(prices)):
        total = 0.0
        ok = True
        for j in range(i - period + 1, i + 1):
            if not math.isfinite(prices[j]):
                ok = False
                break
            total += prices[j]
        if ok:
            out[i] = total / period
    return out


def _rolling_std(prices: list[float], period: int, means: list[float]) -> list[float]:
    out = _nan(len(prices))
    for i in range(period - 1, len(prices)):
        mean = means[i]
        acc = 0.0
        for j in range(i - period + 1, i + 1):
            acc += (prices[j] - mean) ** 2
        out[i] = math.sqrt(acc / period)
    return out


def rsi(prices: list[float], period: int) -> list[float]:
    out = _nan(len(prices))
    if len(prices) <= period:
        return out
    avg_gain = 0.0
    avg_loss = 0.0
    for i in range(1, period + 1):
        change = prices[i] - prices[i - 1]
        if change >= 0:
            avg_gain += change
        else:
            avg_loss -= change
    avg_gain /= period
    avg_loss /= period
    out[period] = 100 if avg_loss == 0 else 100 - 100 / (1 + avg_gain / avg_loss)
    for i in range(period + 1, len(prices)):
        change = prices[i] - prices[i - 1]
        gain = change if change > 0 else 0
        loss = -change if change < 0 else 0
        avg_gain = (avg_gain * (period - 1) + gain) / period
        avg_loss = (avg_loss * (period - 1) + loss) / period
        out[i] = 100 if avg_loss == 0 else 100 - 100 / (1 + avg_gain / avg_loss)
    return out


def atr(s: dict, period: int) -> list[float]:
    high, low, close = s["high"], s["low"], s["close"]
    n = len(close)
    out = _nan(n)
    if n <= period:
        return out
    tr = [0.0] * n
    tr[0] = high[0] - low[0]
    for i in range(1, n):
        tr[i] = max(
            high[i] - low[i],
            abs(high[i] - close[i - 1]),
            abs(low[i] - close[i - 1]),
        )
    total = 0.0
    for i in range(1, period + 1):
        total += tr[i]
    out[period] = total / period
    for i in range(period + 1, n):
        out[i] = (out[i - 1] * (period - 1) + tr[i]) / period
    return out


def adx(s: dict, period: int) -> list[float]:
    """Average Directional Index (Wilder) — trend-strength (0-100; >25 = trending). Added in the
    single-engine (Phase C) era, so worker-only (no TS parity). Uses Wilder smoothing throughout,
    matching the atr() convention above."""
    high, low, close = s["high"], s["low"], s["close"]
    n = len(close)
    out = _nan(n)
    if n <= 2 * period:  # need `period` bars to seed the DIs, then `period` more to seed the ADX
        return out

    tr = [0.0] * n
    plus_dm = [0.0] * n
    minus_dm = [0.0] * n
    for i in range(1, n):
        up = high[i] - high[i - 1]
        down = low[i - 1] - low[i]
        plus_dm[i] = up if (up > down and up > 0) else 0.0
        minus_dm[i] = down if (down > up and down > 0) else 0.0
        tr[i] = max(high[i] - low[i], abs(high[i] - close[i - 1]), abs(low[i] - close[i - 1]))

    # Wilder-smoothed TR / +DM / -DM (seed = sum over the first `period` bars starting at index 1).
    atr_s = [0.0] * n
    pdm_s = [0.0] * n
    mdm_s = [0.0] * n
    atr_s[period] = sum(tr[1:period + 1])
    pdm_s[period] = sum(plus_dm[1:period + 1])
    mdm_s[period] = sum(minus_dm[1:period + 1])
    for i in range(period + 1, n):
        atr_s[i] = atr_s[i - 1] - atr_s[i - 1] / period + tr[i]
        pdm_s[i] = pdm_s[i - 1] - pdm_s[i - 1] / period + plus_dm[i]
        mdm_s[i] = mdm_s[i - 1] - mdm_s[i - 1] / period + minus_dm[i]

    dx = [0.0] * n
    for i in range(period, n):
        if atr_s[i] == 0:
            continue
        plus_di = 100 * pdm_s[i] / atr_s[i]
        minus_di = 100 * mdm_s[i] / atr_s[i]
        denom = plus_di + minus_di
        dx[i] = 100 * abs(plus_di - minus_di) / denom if denom != 0 else 0.0

    # ADX = Wilder-smoothed DX; first value at index 2*period-1 = mean of DX[period .. 2*period-1].
    out[2 * period - 1] = sum(dx[period:2 * period]) / period
    for i in range(2 * period, n):
        out[i] = (out[i - 1] * (period - 1) + dx[i]) / period
    return out


def macd(prices: list[float], fast: int, slow: int, signal_period: int) -> dict:
    ema_fast = ema(prices, fast)
    ema_slow = ema(prices, slow)
    line = [ema_fast[i] - ema_slow[i] for i in range(len(prices))]
    signal = ema(line, signal_period)
    hist = [line[i] - signal[i] for i in range(len(line))]
    return {"line": line, "signal": signal, "hist": hist}


def bollinger(prices: list[float], period: int, mult: float) -> dict:
    mid = sma(prices, period)
    sd = _rolling_std(prices, period, mid)
    upper = [mid[i] + mult * sd[i] for i in range(len(mid))]
    lower = [mid[i] - mult * sd[i] for i in range(len(mid))]
    return {"upper": upper, "mid": mid, "lower": lower}


def rolling_max(values: list[float], period: int) -> list[float]:
    out = _nan(len(values))
    for i in range(period - 1, len(values)):
        m = -math.inf
        for j in range(i - period + 1, i + 1):
            if values[j] > m:
                m = values[j]
        out[i] = m
    return out


def rolling_min(values: list[float], period: int) -> list[float]:
    out = _nan(len(values))
    for i in range(period - 1, len(values)):
        m = math.inf
        for j in range(i - period + 1, i + 1):
            if values[j] < m:
                m = values[j]
        out[i] = m
    return out


def donchian(s: dict, period: int) -> dict:
    return {"upper": rolling_max(s["high"], period), "lower": rolling_min(s["low"], period)}


def stochastic(s: dict, period: int, smooth_k: int, smooth_d: int) -> dict:
    close = s["close"]
    n = len(close)
    hh = rolling_max(s["high"], period)
    ll = rolling_min(s["low"], period)
    raw_k = _nan(n)
    for i in range(period - 1, n):
        rng = hh[i] - ll[i]
        raw_k[i] = 0 if rng == 0 else ((close[i] - ll[i]) / rng) * 100
    k = sma(raw_k, smooth_k)
    d = sma(k, smooth_d)
    return {"k": k, "d": d}
